juju/client/connector.py

```diff
@@ -73,6 +73,12 @@ async def connect(self, **kwargs):
             assert self._connection
             self._log_connection = await Connection.connect(**kwargs)
         else:
+            # TODO (cderici): we need to investigate how to reuse/share
+            # connections between Model & Controller objects
+            # At least either avoid overwriting self._connection with
+            # different types of connections (model, controller), or
+            # have an indication of which type of entity this is
+            # connected to.
```

Comment:
Do we want to do something like refcounted connections that only close when the last reference goes away?

Does the comment about "have an indication" still apply, given that you changed disconnect() to take `entity`?

Reply:
> Does the comment about "have an indication" still apply, given that you changed disconnect() to take `entity`?

No, because you might wanna discover the kind of connection in other places too, depends on what you wanna do, for example we'd get different facades from a model connection then a controller connection. 

So potentially I'd have `self.model_connection` and `self.controller_connection` separately, where we could have a better control which would enable us to avoid doing awkward things like `Connector.connect_model(_model_name)` [first connects to the controller](https://github.com/juju/python-libjuju/blob/54c49e2b704fb67a5881d3d18a935066f78a730a/juju/client/connector.py#L152) to get some info like model uuid etc, [then throws away its connection](https://github.com/juju/python-libjuju/blob/54c49e2b704fb67a5881d3d18a935066f78a730a/juju/client/connector.py#L77) to connect to the model.

> Do we want to do something like refcounted connections that only close when the last reference goes away?

I'd want to move on the opposite direction, actually, to cache the connections and keep it alive in a nice way as long as the process is still going on (and the entity is still alive on the other end). Because a lot of pylibjuju users are still doing the practice in their scripts where they do "connect -> do something -> disconnect -> connect again -> do something else -> disconnect ...". Another solution to this would be to expose nice contexts ([we actually use some internally](https://github.com/juju/python-libjuju/blob/54c49e2b704fb67a5881d3d18a935066f78a730a/tests/base.py#L66)) where they could do:

```python
with model.Model() as m:
  await m.deploy()
```

rather than the current way of doing it like:

```python
m = model.Model()
await m.connect()
await m.deploy()
await m.disconnect()
```

Reply:
I think having a pattern like this could be ok:
```
active_connection = ...
  with model.Model(active_connection, name) as m:
    do stuff

  with model.Model(active_connection, other_name) as m:
    do other stuff
```
or
```
   with active_connection.Model() as m:
     ...
```

But I wouldn't want to have a global (module/package level) cache of active connections that get reused behind your back. (what happens when one breaks, and who was using it, how do you force a reconnection, etc.)